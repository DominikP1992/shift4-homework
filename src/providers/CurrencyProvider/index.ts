export {
  CurrencyProvider,
  useCurrency,
  defaultCurrency,
} from './CurrencyProvider';
