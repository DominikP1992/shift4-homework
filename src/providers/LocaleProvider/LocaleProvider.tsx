import React, { ReactElement, createContext, useContext } from 'react';

export const defaultLocale = 'en-US';
export const LocaleContext = createContext<string | undefined>(undefined);

export const LocaleProvider: React.FC<{
  children: ReactElement | ReactElement[];
}> = ({ children }) => {
  return (
    <LocaleContext.Provider value={defaultLocale}>
      {children}
    </LocaleContext.Provider>
  );
};

export function useLocale() {
  const locale = useContext(LocaleContext);

  if (locale === undefined) {
    throw new Error('useLocale must be used within a LocaleProvider');
  }

  return locale;
}
