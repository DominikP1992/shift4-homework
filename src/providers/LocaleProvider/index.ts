export { useLocale, LocaleProvider, defaultLocale } from './LocaleProvider';
