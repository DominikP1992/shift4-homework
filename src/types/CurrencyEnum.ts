export enum CurrencyEnum {
  USD = 'USD',
  BTC = 'BTC',
  GBP = 'GBP',
  EUR = 'EUR',
  YEN = 'YEN',
}

export type CurrencyType = keyof typeof CurrencyEnum;
