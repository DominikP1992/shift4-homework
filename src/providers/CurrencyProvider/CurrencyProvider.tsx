import { CurrencyEnum, CurrencyType } from '@/types/CurrencyEnum';
import React, { ReactNode, createContext, useContext } from 'react';

export const defaultCurrency = CurrencyEnum.USD;
export const CurrencyContext = createContext<CurrencyType | undefined>(
  undefined,
);

export const CurrencyProvider: React.FC<{
  children: ReactNode | ReactNode[];
}> = ({ children }) => {
  return (
    <CurrencyContext.Provider value={defaultCurrency}>
      {children}
    </CurrencyContext.Provider>
  );
};

export function useCurrency() {
  const currency = useContext(CurrencyContext);

  if (currency === undefined) {
    throw new Error('useCurrency must be used within a CurrencyProvider');
  }

  return currency;
}
