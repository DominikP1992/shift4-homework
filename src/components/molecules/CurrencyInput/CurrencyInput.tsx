import Input from '@/components/atoms/Input';
import { defaultCurrency, useCurrency } from '@/providers/CurrencyProvider';
import { useLocale } from '@/providers/LocaleProvider/LocaleProvider';
import { CurrencyType } from '@/types/CurrencyEnum';
import getSeparators from '@/utils/getSeparators';
import styled from '@xstyled/emotion';
import { useMemo } from 'react';
import { NumericFormat, NumericFormatProps } from 'react-number-format';

export interface CurrencyInputProps
  extends React.InputHTMLAttributes<HTMLInputElement> {
  currency?: CurrencyType;
  fullWidth?: boolean;
}

const InputWithIcon = styled(Input)`
  padding-left: 40px;
  background-image: ${({ currency = defaultCurrency }: CurrencyInputProps) =>
    `url('/icons/${currency.toLowerCase()}.svg')`};
  background-repeat: no-repeat;
  background-position: 10px 50%;
  background-size: 24px;
`;

export default function CurrencyInput({
  currency,
  ...props
}: NumericFormatProps<CurrencyInputProps>) {
  const globalCurrency = useCurrency();
  const locale = useLocale();
  const separators = useMemo(() => {
    return getSeparators(locale);
  }, [locale]);

  return (
    <NumericFormat
      customInput={InputWithIcon}
      thousandSeparator={separators.thousandSeparator}
      decimalSeparator={separators.decimalSeparator}
      currency={currency || globalCurrency}
      allowNegative={false}
      {...props}
    />
  );
}
