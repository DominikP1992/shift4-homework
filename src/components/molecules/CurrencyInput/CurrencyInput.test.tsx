import React from 'react';
import { render, screen } from '@testing-library/react';
import CurrencyInput from './CurrencyInput';
import { useCurrency } from '@/providers/CurrencyProvider';
import { useLocale } from '@/providers/LocaleProvider/LocaleProvider';
import { CurrencyEnum } from '@/types/CurrencyEnum';

jest.mock('@/providers/CurrencyProvider', () => ({
  useCurrency: jest.fn().mockReturnValue('USD'),
}));

jest.mock('@/providers/LocaleProvider/LocaleProvider', () => ({
  useLocale: jest.fn().mockReturnValue('en-US'),
}));

describe('Input', () => {
  beforeEach(() => {
    (useCurrency as jest.Mock).mockReturnValue('USD');
    (useLocale as jest.Mock).mockReturnValue('en-US');
  });

  it('renders without crashing', () => {
    render(
      <CurrencyInput aria-label="test input" currency="USD" value={10000} />,
    );
    expect(screen.getByLabelText('test input')).toBeInTheDocument();
  });

  it('match snapshot', () => {
    render(<CurrencyInput aria-label="test input" value={10000} />);
    expect(screen.getByLabelText('test input')).toMatchSnapshot();
  });

  it('render formatted currency for default (en-us) locale', () => {
    render(<CurrencyInput aria-label="test input" value={10000.25} />);
    expect(screen.getByLabelText('test input')).toHaveValue('10,000.25');
  });

  it('render formatted currency for fr-FR locale', () => {
    (useLocale as jest.Mock).mockReturnValue('fr-FR');

    render(<CurrencyInput aria-label="test input" value={10000.25} />);
    expect(screen.getByLabelText('test input')).toHaveValue('10â€¯000,25');
  });

  for (const currency of Object.keys(CurrencyEnum)) {
    it(`renders proper currency symbol for ${currency}`, () => {
      (useCurrency as jest.Mock).mockReturnValueOnce(currency);
      render(<CurrencyInput aria-label="test input" value={10000.25} />);
      expect(screen.getByLabelText('test input')).toHaveStyle(
        `background-image: url(/icons/${currency.toLowerCase()}.svg)`,
      );
    });
  }
});
