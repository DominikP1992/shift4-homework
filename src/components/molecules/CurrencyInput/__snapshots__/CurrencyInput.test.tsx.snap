// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input match snapshot 1`] = `
<input
  aria-label="test input"
  class="css-4gfppt"
  inputmode="numeric"
  type="text"
  value="10,000"
/>
`;
